main/src/ca/uwaterloo/flix/language/phase/jvm/GenNamespaces.scala

```diff
@@ -17,14 +17,97 @@
 package ca.uwaterloo.flix.language.phase.jvm
 
 import ca.uwaterloo.flix.api.Flix
-import ca.uwaterloo.flix.language.ast.ExecutableAst.Root
-import ca.uwaterloo.flix.language.ast.Type
+import ca.uwaterloo.flix.language.ast.{ExecutableAst, Type}
+import org.objectweb.asm.ClassWriter
+import org.objectweb.asm.Opcodes._
+import ca.uwaterloo.flix.language.ast.Symbol
 
 object GenNamespaces {
 
   // TODO: Documentation and signature
-  def gen(ts: Set[Type], root: Root)(implicit flix: Flix): Map[JvmName, JvmClass] = {
-    Map.empty // TODO
+  def gen(defns: Map[Symbol.DefnSym, ExecutableAst.Def])(implicit flix: Flix): Map[JvmName, JvmClass] = {
+
+    // Defns grouped by their path
+    val groupedDefns = defns.groupBy{ case (sym, defn) =>
+      sym.prefix
+    }
+
+    // Generate each of the namespaces
+    groupedDefns.map{ case (prefix, defnMap) =>
+      val namespace = genNameSpace(prefix, defnMap.values.toList)
+      namespace.name -> namespace
+    }.toMap
+  }
+
+  def genNameSpace(prefix: List[String], defns: List[ExecutableAst.Def])(implicit flix: Flix): JvmClass = {
+
+    // JvmType for namespace
+    val namespace = JvmOps.getNamespaceType(prefix)
+
+    // Class visitor
+    val visitor = new ClassWriter(ClassWriter.COMPUTE_FRAMES) {
+      override def getCommonSuperClass(tpe1: String, tpe2: String): String = {
+        JvmType.Obj.name.toInternalName
+      }
+    }
+
+    // Class header
+    visitor.visit(JvmOps.JavaVersion, ACC_PUBLIC + ACC_FINAL, namespace.name.toInternalName, null,
+      JvmType.Obj.name.toInternalName, null)
+
+    // Adding fields for functional interfaces
+    defns.foreach{ defn =>
+
+      // JvmType of the `tpe` //TODO: wrong type, needs to be the type of the class, not the interface
```

Comment:
It **is** okay to do this:

```
class MyNs {
  Fn$2XYZ foo = new Def$foo
}
```

and I think that is what you are doing.

Reply:
Yeah, but I still need to have the type for `Def$foo`